$MirOS: ports/net/djbdns/patches/patch-src_rbldns-data_c,v 1.1 2005/12/14 23:28:37 tg Exp $
--- src/rbldns-data.c.orig	Sun Feb 11 21:11:23 2001
+++ src/rbldns-data.c	Tue Jan 17 20:53:13 2006
@@ -42,7 +42,7 @@ void syntaxerror(const char *why)
 }
 void die_datatmp(void)
 {
-  strerr_die2sys(111,FATAL,"unable to create data.tmp: ");
+  strerr_die2sys(111,FATAL,"unable to create data.cdb.tmp: ");
 }
 
 int main()
@@ -59,7 +59,7 @@ int main()
   if (fd == -1) strerr_die2sys(111,FATAL,"unable to open data: ");
   buffer_init(&b,buffer_unixread,fd,bspace,sizeof bspace);
 
-  fdcdb = open_trunc("data.tmp");
+  fdcdb = open_trunc("data.cdb.tmp");
   if (fdcdb == -1) die_datatmp();
   if (cdb_make_start(&cdb,fdcdb) == -1) die_datatmp();
 
@@ -89,6 +89,15 @@ int main()
         if (cdb_make_add(&cdb,"",0,tmp.s,tmp.len) == -1)
           die_datatmp();
         break;
+      case '!': /* root entry */
+	j = byte_chr(line.s + 1,line.len - 1,':');
+	if (j >= line.len - 1) syntaxerror(": missing colon");
+	if (ip4_scan(line.s + 1,ip) != j) syntaxerror(": malformed IP address");
+	if (!stralloc_copyb(&tmp,ip,4)) nomem();
+	if (!stralloc_catb(&tmp,line.s + j + 2,line.len - j - 2)) nomem();
+        if (cdb_make_add(&cdb,"R",1,tmp.s,tmp.len) == -1)
+          die_datatmp();
+        break;
       case '0': case '1': case '2': case '3': case '4':
       case '5': case '6': case '7': case '8': case '9':
 	if (!stralloc_0(&line)) nomem();
@@ -121,8 +130,8 @@ int main()
   if (cdb_make_finish(&cdb) == -1) die_datatmp();
   if (fsync(fdcdb) == -1) die_datatmp();
   if (close(fdcdb) == -1) die_datatmp(); /* NFS stupidity */
-  if (rename("data.tmp","data.cdb") == -1)
-    strerr_die2sys(111,FATAL,"unable to move data.tmp to data.cdb: ");
+  if (rename("data.cdb.tmp","data.cdb") == -1)
+    strerr_die2sys(111,FATAL,"unable to move data.cdb.tmp to data.cdb: ");
 
   _exit(0);
 }
